[BME280] Allow to take samples using taskRun (#3879)

TD-er opened the pull request:
Fixes: #3879

TD-er committed:
[BME280] Allow to take samples using taskRun (#3879)

Fixes: #3879

chemmex commented:
TaskRun part seems to be Ok. And the values come, as promised, within 2 seconds

```
2348978: HTTP: taskrun,BME
2348979: Command: taskrun
2349322: BME280: dew point 13.86C
2349333: BME280 : Address: 0x76
2349333: BME280 : Temperature: 25.68
2349334: BME280 : Humidity: 48.07
2349334: BME280 : Barometric Pressure: 972.99
2349338: EVENT: BME#All=25.68,48.07,972.99
```

TD-er commented:
Great!
Thanks for testing.
I was cleaning up the code a bit more, but had not yet come to test it.

TD-er committed:
[BME280] Clean-up code

TD-er committed:
[BME280] Proper handling error states, send error event.

The BME280 plugin does not always return true on PLUGIN_READ as it first starts the measurement and then reschedules itself.

This may lead to constantly getting error state values, only to be updated after a successful measurement.

TD-er committed:
[BME280] Make TaskError event more useful by adding error string

tonhuisman commented on src/_P028_BME280.ino:
```diff
@@ -126,9 +129,9 @@ boolean Plugin_028(uint8_t function, struct EventStruct *event, String& string)
 
       if (nullptr != P028_data) {
-        if (P028_data->sensorID != Unknown_DEVICE) {
+        if (P028_data->sensorID != P028_data_struct::Unknown_DEVICE) {
```
Doesn't PVS Studio suggest to combine these `if` statements? Or do we leave that to the optimization phase of the compiler?

TD-er replied:
Nested if statements will be optimized by the compiler.
What PVS Studio does (often) suggest is to make complex compares simpler.
Especially if the same boolean value is used on opposite sides of a logic operator.

Something like these:
```
if (A || (!A && B))
```

Very simple to do optimizations will be taken care of by the compiler.
```
if (A) {
  if (B) {
    ...
  }
}
```
Will result in the same compiled output as this:
```
if (A && B) {
  ...
}
```

Only exception might be when there are some variables defined within the inner scope.

tonhuisman commented on src/src/DataTypes/ESPEasy_plugin_functions.h:
```diff
@@ -49,7 +49,7 @@
 #define PLUGIN_WEBFORM_SHOW_ERRORSTATE_OPT 42 // Show Error State Value options, so be saved during PLUGIN_WEBFORM_SAVE
 #define PLUGIN_INIT_VALUE_RANGES           43 // Initialize the ranges of values, called just before PLUGIN_INIT
-#define PLUGIN_GET_ERROR_VALUE_STATE       44 // Should we return the error values or not, called when PLUGIN_READ returns false, must also set error values into UserVar[]
+#define PLUGIN_READ_ERROR_OCCURED          44 // Function returns "true" when last measurement was an error., called when PLUGIN_READ returns false, must also set error values into UserVar[]
```
Does this comment still state the correct/expected behavior? (the last part, about filling the error values)

TD-er replied:
Nope, you're right the comment needs to be updated.

TD-er committed:
[BME280] Update comment to reflect actual behaviour

TD-er committed:
[Cleanup] Reduce build size web form functions

TD-er committed:
[Cleanup] Reduce bin size deduplicate webform switch plugins (001/009/019)

TD-er commented:
Added some more cleanup, so now the ESP8266 testing D builds may also fit again.

TD-er committed:
[Cleanup] Uncrustify

TD-er committed:
Merge branch 'mega' into bugfix/BME280

TD-er committed:
[Cleanup] Minor optimizations to reduce build size

TD-er committed:
[Static Files] Fix option to local store dev page JS file (#4102)

Fixes: #4102

tonhuisman commented on src/src/WebServer/Markup_Forms.h:
```diff
@@ -109,7 +121,19 @@ void addFormFloatNumberBox(const String& label,
+                           float         min,
+                           float         max,
+                           uint8_t       nrDecimals = 6,
+                           float         stepsize   = 0.0f
```
Wouldn't it be more useful to have a step-size of 1.0f ?

TD-er replied:
Nope...
See:
```c++
  if (stepsize <= 0.0f) {
    html += F("0.");

    for (uint8_t i = 1; i < nrDecimals; ++i) {
      html += '0';
    }
    html += '1';
  } else {
    html += toString(stepsize, nrDecimals);
  }
```
In other words, by not setting the stepsize, you will get one based on the nr of decimals..

Not sure if that's practical though.

TD-er committed:
[Build] Show proper ESP board name on ESP8266 sysinfo

Apparently the generated board name was done different for ESP8266 compared to ESP32.
Now both platforms show the name as given in the board definition JSON file

TD-er committed:
[Build] Do not use commas in ESP board name
